lib/pure/asyncdispatch.nim

```diff
@@ -1093,31 +1146,90 @@ else:
       writeList: seq[Callback]
 
-    AsyncEvent* = distinct SelectEvent
+    AsyncEventImpl = object
+      triggered: bool
+      when compileOption("threads"):
+        eventLock: RLock
+      p: PDispatcher
+      vFD: VirtualFD
+      callbacks: seq[Callback]
+    AsyncEvent* = ptr AsyncEventImpl
+    #AsyncEvent* = distinct SelectEvent
 
     PDispatcher* = ref object of PDispatcherBase
       selector: Selector[AsyncData]
+      virtualHandles: Table[VirtualFD, AsyncEvent] # pseudo handles for custom AsyncEvents.
+      nextVirtualHandle: VirtualFD
+      virtualMuxHandle: SelectEvent # all the virtual handles get multiplexed through a single handle.
 
   proc `==`*(x, y: AsyncFD): bool {.borrow.}
-  proc `==`*(x, y: AsyncEvent): bool {.borrow.}
+  proc `==`*(x: VirtualFD, y: VirtualFD): bool {.borrow.}
+  proc hash(x: VirtualFD): Hash {.borrow.}
+
+  proc `==`*(x, y: AsyncEvent): bool = x.vFD == y.vFD
+
+  const InvalidVirtualFD = VirtualFD(-1)
 
   template newAsyncData(): AsyncData =
     AsyncData(
       readList: newSeqOfCap[Callback](InitCallbackListSize),
       writeList: newSeqOfCap[Callback](InitCallbackListSize)
     )
 
+  proc unregister*(ev: AsyncEvent) =
+    ## Unregisters event ``ev``.
+    doAssert(ev.vFD != InvalidVirtualFD, "Event is not registered in the queue!")
+    withLockIfThreads ev.eventLock:
+      ev.p.virtualHandles.del ev.vFD
+      ev.p = nil
+      ev.vFD = InvalidVirtualFD
+
   proc newDispatcher*(): owned(PDispatcher) =
     new result
     result.selector = newSelector[AsyncData]()
     result.timers.newHeapQueue()
+    result.virtualHandles = initTable[VirtualFD, AsyncEvent]()
+    result.timers.newHeapQueue()
     result.callbacks = initDeque[proc ()](InitDelayedCallbackListSize)
+    result.virtualMuxHandle = newSelectEvent()
+
+    # register the real listener for all the virtual events
+    var p = result
+    proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+      # find the virtual events that triggered the physical event and
+      # add them to the callback list.
+      # Not very efficient, but requires the least coordination between threads.
+      
+      proc vcbFactory(vfd: VirtualFD, cbIndex: int) : proc() {.closure, gcsafe.} =
+        let ev = p.virtualHandles[vfd]
+        result = proc() {.closure, gcSafe.} =
+          let removeVCB = ev.callbacks[cbIndex](vfd.AsyncFD)
+          if removeVCB:
+            withLockIfThreads ev.eventLock:
+              ev.callbacks.del(cbIndex)
+              if ev.callbacks.len == 0:
+                ev.unregister
+
+      for vfd, ev in p.virtualHandles:
+        withLockIfThreads ev.eventLock:
+          if ev.triggered:
+            ev.triggered = false
+            for i in 0..ev.callbacks.high:
+              p.callbacks.addLast(vcbFactory(vfd, i))
+
+      return false 
+
+    var data = newAsyncData()
+    data.readList.add(virtualMuxCB)
+    result.selector.registerEvent(result.virtualMuxHandle, data)
```

Comment:
This sort of code would do well to be put in a separate procedure called ``initVirtualAsyncEvents`` or the like.